### PR TITLE
refactor(admin): Migrate Batch Export Feature

### LINKED ISSUE #3451: VxAdmin: move batch tally csv export to backend

### PR DESCRIPTION
**Pull request by adghayes**
_review commit by commit_

## Overview

Closes #3451. Moves the batch tally export to the backend. Batch tally is computed with the new tally infrastructure, its first application 🎉. I had to add a new "Save Modal" because the other one uses `kiosk-browser` to save the file - the new modal expects a `Result` of saving the file on the backend.

Must be reviewed and merged with https://github.com/votingworks/kiosk-browser/pull/149.

## Demo Video or Screenshot

https://github.com/votingworks/vxsuite/assets/37960853/ab8d72ca-bac5-4fbe-af28-37997001e9cc

## Testing Plan

- Automated Testing
- Manual Testing

## Checklist

- [x] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.

- [x] I have added a screenshot and/or video to this PR to demo the change
- [ ] ~~I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates~~

### CODE REVIEW

File: apps/admin/frontend/src/components/save_file_modal.tsx
Comment (adghayes) [inline apps/admin/frontend/src/components/save_file_modal.tsx:82]: This component is largely copied from `save_file_to_usb` but simplified where possible.

File: apps/admin/frontend/src/components/save_file_modal.tsx
Comment (adghayes) [inline apps/admin/frontend/src/components/save_file_modal.tsx:121]: All of these errors are unexpected and would be the equivalent of 500 errors, so not handling specifically.

File: apps/admin/backend/package.json
Comment (adghayes) [inline apps/admin/backend/package.json:52]: @benadida had commented on an original PR for the CSV exports that we should probably be using a package to handle all the CSV formatting rather than trying to do it ourselves, so here I introduce the most common and supported one.

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (kofi-q) [inline apps/admin/frontend/src/components/save_backend_file_modal.tsx:41]: Is this intended to be used outside of this file? If so, can we break it out to a separate component file?

File: apps/admin/frontend/src/components/export_batch_tally_results_button.tsx
Comment (kofi-q) [inline apps/admin/frontend/src/components/export_batch_tally_results_button.tsx:27]: If it's handled upstream, do we even need a `catch` here?

**Review by kofi-q — COMMENTED** [review]
Still reviewing, but need to check out for the day - got through the frontend changes and left a few comments. Will finish up in the morning - sorry for blocking!

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx:143]: 748482989

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx:173]: The `if` block is just to place a check on `currentState` to ensure we've handled all the values appropriately, similar to how we would in the `default` of a `switch-case`. It's a compile-time check that could be removed if it is too confusing.

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx:40]: The intention is, for now, no. I thought about the possibility of using it later, but it could be broken out when that time comes.

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx:107]: I originally stripped it because I thought it unnecessary but prefer the UX with a bit of a pause. Added a comment in adb1b9434:
```
    // Wait at least 1 second before changing the state to "done" or "error". Otherwise
    // the "Saving…" message might flash very quickly on the screen, giving the user the
    // impression they missed something.
```

File: apps/admin/frontend/src/components/export_batch_tally_results_button.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/export_batch_tally_results_button.tsx:25]: This return value of this callback is used by `SaveBackendFileModal` to determine whether to go to a success or error state, so I think it does need to be returned.

File: apps/admin/frontend/src/components/export_batch_tally_results_button.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/export_batch_tally_results_button.tsx:27]: Great question. To be honest, I do not understand the react query error handling, but this is how I've learned from @arsalansufi and @jonahkagan to handle errors in cases where I have to do `async` mutations. I know it's clunky. See the [rearchitecture presentation](https://docs.google.com/presentation/d/1qcohhkeavCVjTZAAhGbbB7sbnvDJLlNov3OTLijfcvs/edit#slide=id.g1f8414d696e_0_361).

File: apps/admin/frontend/src/components/export_batch_tally_results_button.tsx
Comment (kofi-q) [reply apps/admin/frontend/src/components/export_batch_tally_results_button.tsx:25]: Oops sorry, meant to delete this comment - I'd confused this with the `onSave` prop for `SaveAsButton` when I scrolled down to check

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (kofi-q) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx:175]: Ah, that makes sense. Yeah, a `switch-case` might be cleaner/more readable then, but this minor, so your call

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (kofi-q) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx]: Hm, not a big fan of that as a pattern - I think the fact that we have confirmation saying the save was successful should be enough signal to the user. Unless it's solving a user problem we've seen before - even then, I'd wonder if maybe the confirmation needs to be more obvious instead?

**Review by kofi-q — APPROVED** [review]
Got through backend pieces - LGTM!

Some code pattern concerns in the frontend pieces, but consider non-blocking

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx]: b00d3b2c0

File: apps/admin/frontend/src/components/save_backend_file_modal.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx]: Yeah, I think one of the main problems is that the confirmation is not very obvious. It's not using any color or iconography, so you just have to notice that the text changed.

Comment (adghayes) [reply apps/admin/frontend/src/components/save_backend_file_modal.tsx]: I just removed the delay in d2b5fc6cf, but going to punt on improving the confirmation.

File: apps/admin/frontend/src/components/export_batch_tally_results_button.tsx
Comment (adghayes) [reply apps/admin/frontend/src/components/export_batch_tally_results_button.tsx]: https://github.com/votingworks/vxsuite/pull/3573/commits/bc7247235488518bb5a6c533678b585bdcb2929c

**Review by kofi-q — APPROVED** [review]
Refactor looks good! Liking the focused, state-based tests too 👌

**Pull request merged**
